### ISSUE TITLE
Chunk cache gets dropped/deleted if dataset is not aliased

### ISSUE THREAD

**Issue by gpsgibb** [h5py/h5py#1960]
## OS/h5py version info

PopOS 21.04
h5py    3.4.0
HDF5    1.12.1
Python  3.9.5 (default, May 11 2021, 08:20:37) 
[GCC 10.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.19.5
cython (built with) 0.29.24
numpy (built against) 1.19.3
HDF5 (built against) 1.12.1

Also occurs on:
CentOS Linux release 7.6.1810 (Core)
h5py    3.1.0
HDF5    1.10.6
Python  3.6.7 | packaged by conda-forge | (default, Mar  4 2020, 16:55:12) 
[GCC 7.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.17.2
cython (built with) 0.29.21
numpy (built against) 1.16.5
HDF5 (built against) 1.10.6

## Problem description

If I read a slice from a chunked dataset, the chunk cache only seems to persist if I alias the dataset first, and read from the alias. If I access the dataset directly (without aliasing) it does not seem to retain the cache e.g.:
```python
with h5py.File(filename,"r") as f:
    # chunks are cached
    dataset = f[dataset_name]
    slice = dataset[...]

    # chunks are not cached
    slice = f[dataset_name][...]
```

It seems that every time the dataset is accessed by `f[dataset_name]` the chunk cache is dropped or cleared. Possibly python's garbage collection deleting the cache every time `f[dataset_name]` goes out of scope?

I'm unsure as to whether this is a bug, or more a feature of how h5py works, although given that one specifies the chunk cache size upon opening the file, it's odd that the cache here seems to live with the dataset instance rather than with the file object itself, or at least attached to the file object. Should this behaviour be mentioned in the caching documentation for clarity?

## Working example

Code to write the dataset we will be reading:
```python
import h5py
import numpy as np

#create file with a single chunked dataset
with h5py.File("testfile.h5","w") as f:
    
    #create some dummy data
    data = np.zeros((4096,4096))

    dataset = f.create_dataset(name="test",data=data,chunks=(128,128))
```

Code that demonstrates correct caching (`good.py`):
```python
import h5py

with h5py.File("testfile.h5","r",rdcc_nbytes=(1024*1024*10)) as f:
    #alias the dataset "test" to the variable "test"
    test = f["test"]
    #read the same 400x400 square 100 times
    for i in range(100):
        x = test[100:500,100:500]
```

Code that demonstrates lack of caching (`bad.py`):
```python
import h5py

with h5py.File("testfile.h5","r",rdcc_nbytes=(1024*1024*10)) as f:
    #read the same 400x400 square 100 times
    for i in range(100):
        x = f["test"][100:500,100:500]
```

In the test code, I repeatedly extract the same 400x400 slice from the dataset. The expected behaviour would be for 4 x 4 = 16 128x128 chunks to be read from the file into the cache in the first iteration, then no more reads from the file for the subsequent iterations as all the required data should be stored in the cache. 

Examining the system calls when executing `good.py`, (e.g. with`strace`) there are, as expected, 16 calls to `pread64` which read in 131072 bytes each, corresponding to a 128x128x8 bytes chunk. Examining the system calls for `bad.py`, however, there are 1600 such reads, demonstrating that here it reads the 16 chunks from the file every time the dataset is sliced into (100x) and the chunk cache appears not to be retained between iterations.

**Comment by tacaswell** [comment]
This is an amazingly useful bug report!  That said I think your analysis is mostly correct and there is nothing we can (or should) do about this other than document it thoroughly. 

-----

> Possibly python's garbage collection deleting the cache every time f[dataset_name] goes out of scope?

I suspect that this is correct in practice, but propose a different mechanism.   we are clearing the cache by de-allocating the underlying c struct that represents the cache.  My guess is that libhdf is saying "well, if you no longer care about that data set I do not either!" and purging its chunks from the cache.   I took a quick look a quick skim of the libhdf5 code and found a comment where it says "when the reference count goes to 0 the dataset gets deallocated", but could not find where that logic lived, there appears to be a callback system that I did not chase through that fires when their reference count goes to zero.

Similarly, on Python level when you give up all your references to a data set object we let Python tear it down (there is a bunch of complexity to _keep_ groups and files alive if you still have a child from them) and internally we decrease the hdf5 reference (see https://github.com/h5py/h5py/blob/8a22294e27be0678f755d66b288b2d97b258b633/h5py/_objects.pyx#L199 and https://github.com/h5py/h5py/blob/8a22294e27be0678f755d66b288b2d97b258b633/h5py/_objects.pyx#L216).

I suspect that to keep libhdf5 from tearing down the data set and the cache we would have to keep the python-side data set alive for you and then be in the business of sorting out _how long_ to keep that alive for you.

For better or worse, I think expecting the user to keep a reference to data sets they want to re-use is a reasonable position to take.  Even without caching it should save you a bit of time as it save you from re-doing the work of reading all of the meta data about the dataset from disk.

**Comment by gpsgibb** [comment]
To be honest, I have no idea why I didn't keep a reference to the datasets in my application code. It was easy enough to implement and now my code is reading the volume of data from the disk that I expected it to read. TBH performance wasn't affected much by the dropped chunk caches as the portions of the file I was reading repeatedly (hence why I was using the chunk cache) were loaded into disk cache by the OS, but if I allocate chunk cache I want to be using it. I also suspect that not keeping a reference meant that every time I accessed a dataset some init code was executed (nominally to re-establish the cache) so I was better off keeping a reference to the dataset anyway.

**Comment by tacaswell** [comment]
> nominally to re-establish the cache

I suspect that hdf5 is also reading and building a bunch of meta-data about the location of the chunks in the data set and setting up what ever data structures it needs to run the selection logic.

**Comment by takluyver** [comment]
> My guess is that libhdf is saying "well, if you no longer care about that data set I do not either!" and purging its chunks from the cache.

Yes, I'm pretty sure this is the case. `f['dsname']` is equivalent to an `H5Dopen` call in C, and letting the Dataset object go out of scope is equivalent to calling `H5Dclose` (i.e. close a dataset handle). The explicit close & open calls just make it more obvious if you're repeatedly closing & reopening something you could be reusing.

> there is nothing we can (or should) do about this other than document it thoroughly.

I'm inclined to agree. We *could* have a cache on a file object of recently accessed datasets, which we can retrieve without reopening. But I can easily imagine that introducing a bunch of bugs where it intersects with features like external links and SWMR. And even if we avoid those, it's another piece to understand if you're e.g. tracking memory use or debugging a tricky issue.
